src/app/clusters/concentration-measurement-server/concentration-measurement-server.cpp

```diff
@@ -0,0 +1,512 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app/InteractionModelEngine.h>
+#include <app/SafeAttributePersistenceProvider.h>
+#include <app/clusters/concentration-measurement-server/concentration-measurement-server.h>
+#include <app/clusters/on-off-server/on-off-server.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/config.h>
+#include <platform/DiagnosticDataProvider.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ConcentrationMeasurement {
+
+Instance::Instance(EndpointId aEndpointId, ClusterId aClusterId, MeasurementMediumEnum aMeasurementMedium, uint32_t aFeature) :
+    AttributeAccessInterface(Optional<EndpointId>(aEndpointId), aClusterId), mEndpointId(aEndpointId), mClusterId(aClusterId),
+    mMeasurementMedium(aMeasurementMedium), mFeature(aFeature)
+{
+    mMeasuredValue.SetNull();
+    mMinMeasuredValue.SetNull();
+    mMaxMeasuredValue.SetNull();
+    mPeakMeasuredValue.SetNull();
+    mAverageMeasuredValue.SetNull();
+
+    mMeasurementUnit = MeasurementUnitEnum::kUnknownEnumValue;
+    mLevel           = LevelValueEnum::kLow;
+
+    VerifyOrDieWithMsg(Init() == CHIP_NO_ERROR, Zcl, "Concentration Measurement failed to initialize for endpoint:%d", mEndpointId);
+}
+
+Instance::Instance(EndpointId aEndpointId, ClusterId aClusterId, MeasurementMediumEnum aMeasurementMedium,
+                   MeasurementUnitEnum aMeasurementUnit, uint32_t aFeature) :
+    AttributeAccessInterface(Optional<EndpointId>(aEndpointId), aClusterId),
+    mEndpointId(aEndpointId), mClusterId(aClusterId), mMeasurementMedium(aMeasurementMedium), mMeasurementUnit(aMeasurementUnit),
+    mFeature(aFeature)
+{
+    VerifyOrDieWithMsg(HasFeature(Feature::kNumericMeasurement), Zcl,
+                       "For endpoint %d, the numeric measurement feature must be supported if setting a unit", mEndpointId);
+
+    mMeasuredValue.SetNull();
+    mMinMeasuredValue.SetNull();
+    mMaxMeasuredValue.SetNull();
+    mPeakMeasuredValue.SetNull();
+    mAverageMeasuredValue.SetNull();
+
+    mLevel = LevelValueEnum::kLow;
+
+    VerifyOrDieWithMsg(Init() == CHIP_NO_ERROR, Zcl, "Concentration Measurement failed to initialize for endpoint:%d", mEndpointId);
+}
+
+bool Instance::HasFeature(Feature feature) const
+{
+    return (mFeature & to_underlying(feature)) != 0;
+}
+
+CHIP_ERROR Instance::SetMeasuredValue(DataModel::Nullable<float> aMeasuredValue)
+{
+    if (!HasFeature(Feature::kNumericMeasurement))
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+
+    if (!aMeasuredValue.IsNull())
+    {
+        if (!mMaxMeasuredValue.IsNull())
```

Comment:
Also worth thinking about: a helper function that takes (possibly null) min, measured, max and returns whether that's a valid configuration.  That could be used here, in the min/max setters, in `SetPeakMeasuredValue` (which is currently going to crash if min/max are null), etc.